Cycles node updates

BrianHanke opened the pull request:
Hi John, here's what I've come up with so far to update all Cycles nodes to match Blender's layout, plus improve overall usability.

It's all working well in Gaffer as far as I can tell, but I think it could be cleaner for sure. I have some questions/comments too:

1. I tried updating the SSS radius defaults on line 426 but that syntax didn't work. Not sure why. Just did in manually at line 538.
2. I made 4 separate functions starting at line 441. I had the idea to combine them, but not sure if that's possible/desirable.
3. `environment_texture` has a default `tex_mapping__scale.x` value of -1, is that correct?
4. All the texture mapping sections have default min/max values of +/-340282346638528859811704183484516925440. As a user that looks kind of intimidating, was wondering if there was a better way to present that value.
5. The Blender `principled_hair_bsdf` has one of those interfaces that replaces parts of itself depending on what you have selected. I didn't like having a raw list of all the parameters, even the ones not "available", so what I did was put everything into sections. It feels ok, but it doesn't match Blender and I'm not sure it's the best solution.
6. If I was being really perfectionist (I can't help it!) I'd prefer to have the tab menu entries like _Combine Hsv_ read as _Combine HSV_. But that's a pretty minor point...

I think covers it. Looking forward to your thoughts! 
Best,

Brian

boberfly commented on startup/GafferCyclesUI/metadata.py:
```diff
@@ -0,0 +1,592 @@
+
+defaults = {
+    "principled_bsdf" : {
+        #"subsurface_radius.x" : 1, why doesn't this work?
```
I think `"subsurface_radius" : imath.Color3f( 1, .2, .1 )` might be what you're after here.

johnhaddon replied:
That sounds right. The metadata is registered against `parameters.*` which only applies to the immediate children of `parameters`. There is a `parameters...*` syntax you could use to match recursively, but I think @boberfly's suggestion is better.

johnhaddon reviewed (commented):
Thanks Brian, this is a great improvement!

> I tried updating the SSS radius defaults on line 426 but that syntax didn't work. Not sure why. Just did in manually at line 538.

I think @boberfly's suggestion should have you covered there, so we should be able to remove the manual bits.

> I made 4 separate functions starting at line 441. I had the idea to combine them, but not sure if that's possible/desirable.

That does sound like a good idea - it will make future extensions easier. I think the best way of doing it might be to group everything into one dictionary, closer to what we use for registering node metadata, with everything related to one parameter in the same place. Something like this (untested sketch) :

```
parameterMetadata = {

    "principled_bsdf" : {

             "subsurface_method" :  {
                 
                 "label" : "Method",
                 "layout:section" : "Subsurface", 
                 "layout:index" : 8,

              },
              ...
     },
     ...
}

def metadata( plug, name ) :

    global parameterMetadata

    shaderName = plug.node()["name"].getValue()
    shaderDict = parameterMetadata.get( shaderName )
    if shaderDict is None :
        return None

     parameterDict = shaderDict.get( plug.getName() )
     if parameterDict is None :
         return None

    return parameterDict.get( name )

for name in ( "label", "layout:section", "layout:index" ) :
      Gaffer.Metadata.registerValue(
             GafferCycles.CyclesShader, "parameters.*", name, functools.partial( metadata, name = name )
       )
```

> environment_texture has a default tex_mapping__scale.x value of -1, is that correct?

It was definitely intended, and I _think_ it's correct. This was specified in `startup/GafferCycles/userDefaults.py`. I realise that we also have a single metadata registration in `startup/GafferCycles/principledBSDFCompatibility.py`. It would be great if you could consolidate those both into your new file.

> All the texture mapping sections have default min/max values of +/-340282346638528859811704183484516925440. As a user that looks kind of intimidating, was wondering if there was a better way to present that value.

I agree. I don't even know if those parameters are actually useful, or if it's more common to plug something in to `vector` for finer control. I don't see the parameters in Blender at all, but I'm less than a novice in there. If we want to keep them, perhaps one option is to hide them or grey them out unless `use_min_max` is on (see below).

> The Blender principled_hair_bsdf has one of those interfaces that replaces parts of itself depending on what you have selected. I didn't like having a raw list of all the parameters, even the ones not "available", so what I did was put everything into sections. It feels ok, but it doesn't match Blender and I'm not sure it's the best solution.

Gaffer has metadata for this too - `layout:visibilityActivator`. That could be added to the big dict-of-everything (see above) pretty easily. But instead of putting a constant value in the dict, you'd refer to a function which would compute the result on the fly. Like your existing `section()` function, this would be passed the plug, and could therefore find the 
`scattering_mode` plug on the same node, read its value, and then return True or False to control the visibility of the other plugs. Does that make sense?

> If I was being really perfectionist (I can't help it!) I'd prefer to have the tab menu entries like Combine Hsv read as Combine HSV. But that's a pretty minor point...

You'll find the logic for that in python/GafferCyclesUI/ShaderMenu.py, where we choose `displayName` at line 56. A little hack in there might alleviate my CamelCaseTlaOcd too...

Thanks again!

johnhaddon commented:
One other note : we'll ideally target this PR at the `1.4_maintenance` branch so it can go into the next `1.4.x` release (`main` is where we develop what will become `1.5`). But we can deal with that once you're happy with the rest...

BrianHanke commented:
@johnhaddon and @boberfly thanks for the great comments! I'll make some changes over the next few days and post the results.

BrianHanke commented:
Lots of updates! Everything is working like I imagined! The big changes are:

- Added function to show/hide parameters in `principled_hair_bsdf` and `vector_rotate` to match Blender.
- The `tex_mapping` min/max fields are now hidden until _Use Min Max_ is checked.
- Moved defaults to `metadata.py` from `principledBSDFCompatibility.py` and `userDefaults.py`.
- There is now one master dictionary with all parameters.
- Added a few lines to convert all the Cycles menu entries to title case.

There are two small things I'm still not sure about:

- Technically, in `vector_rotate` _Angle_ should be hidden only when `Euler xyz` is selected, but I couldn't figure out how to hide a parameter based on a selection.
- I put all the shared light parameters like _Cast Shadow_ and _Use Diffuse_ under a section called _Contribution_. Is that the right word to describe those items?

johnhaddon commented on startup/GafferCyclesUI/metadata.py:
```diff
@@ -0,0 +1,923 @@
+
+mapping = [ "parameters.tex_mapping__translation", "parameters.tex_mapping__rotation", "parameters.tex_mapping__scale", "parameters.tex_mapping__use_minmax", 
+           "parameters.tex_mapping__min", "parameters.tex_mapping__max", "parameters.tex_mapping__x_mapping", "parameters.tex_mapping__y_mapping", 
+           "parameters.tex_mapping__z_mapping", "parameters.tex_mapping__type", "parameters.tex_mapping__projection" ]
```
Formatting nitpick : please close braces at the same indentation level as they were opened at.

```
mapping = [ 
    ...
]
```

johnhaddon reviewed (commented):
Thanks Brian, this is making the Cycles UIs so much more approachable. I've made a few comments inline, and have answers to you questions below...

> Technically, in vector_rotate Angle should be hidden only when Euler xyz is selected, but I couldn't figure out how to hide a parameter based on a selection.

I've demonstrated an approach to that in this commit : https://github.com/johnhaddon/gaffer/commit/b3444878ad2db8f4f83a3a386d4367459a7cc9ad

I think this is a better approach to the visibility activators generally, and should allow you to remove the `visibility()` function entirely, moving all the activators into the main metadata dict.

> I put all the shared light parameters like Cast Shadow and Use Diffuse under a section called Contribution. Is that the right word to describe those items?

I don't know, but having them tucked away in a section seems preferable to not, so I suggest we run with it for now. The nice thing about all this is that it's just metadata, and we can change our minds later without breaking anybody's scenes. I couldn't find a bunch of those settings in Blender though - are they not exposed there, or am I just looking in the wrong place?

Now we're getting close to merging, there are a few procedural nitpicks it would be good to take care of too, although I can do these for you if you prefer :

- Give your branch a name other than `main`, and rebase it on Gaffer's `1.4_maintenance` branch, including only the commits you want to be merged. Currently it is based on `main`, which will become Gaffer 1.5, and contains commits we don't want to merge for 1.4. There is a brief description of how we manage branches [here](https://github.com/GafferHQ/gaffer/wiki/Branches-and-Versioning), although the GitHub docs will probably be more useful.
- Update Changes.md with a note in the Improvements section. More details about that [here](https://github.com/GafferHQ/gaffer/wiki/Editing-Changes.md).
- Squash the commits into a single one for the metadata changes (including the removals from the other `.py` files) and a single one for the menu name change. If you're not tired by now, have a read of our standards for commit logs [here](https://github.com/GafferHQ/gaffer/wiki/Commit-Message-Format).

Thanks!

BrianHanke commented:
Thanks for the feedback, John. The new visibilityActivator system looks way better than what I came up with. That's some advanced Python I didn't know about! I'll make the other changes over the next couple of days. Two points:

1. I'm using VS Code and I'm hitting tab, but I guess it formats tabs as spaces? Maybe there's a setting to change that.
2. I dug around in Blender and I see now how they set up the light "contribution" stuff:

![blender01](https://github.com/GafferHQ/gaffer/assets/59420805/d6428fcb-c6b5-4738-9146-b3ae7eab3a5b)
![blender02](https://github.com/GafferHQ/gaffer/assets/59420805/7acc6743-0ce7-4544-95be-5d523656da99)

They call it "Ray Visibility" and Cast Shadow, MIS and Shadow Caustics are in a separate place. They also put the light size and shape stuff under a "Beam Shape" section. I'll look into setting things up to match that, although I kind of like how it's laid out now.

johnhaddon commented:
> I'm using VS Code and I'm hitting tab, but I guess it formats tabs as spaces? Maybe there's a setting to change that.

Yeah. I think by default it attempts to follow the convention of the file you're editing, but when making one from scratch there's a setting that probably defaults to spaces.

>  I dug around in Blender and I see now how they set up the light "contribution" stuff:

Coolcool. I didn't mind the way you had it laid out either, but matching Blender probably makes sense at the moment - helps folks find the things they're used to.

BrianHanke commented:
@johnhaddon Hey John, all the changes are done, everything is working perfectly. I'm having GitHub problems though trying to switch branches. I renamed my branch as you suggested and then tried rebasing based on a guide I found, but that gave me a lot of errors. Now I come back here and it said I deleted main and closed this PR. Gah! Would it be easier to start over with a PR from a new fork of 1.4_maintenance?

danieldresser-ie commented:
I don't think opening a new PR is likely to help much - the important thing is for you to get a correctly named branch in the right state - once you've got that, you can edit this PR to point to your freshly created branch.

When you renamed the branch, did you do that using Github web interface, operating on your repo on github ( that would explain why the main branch on the repo this PR was pointing to disappeared, and this PR got closed)? I would usually just rename my branch locally, and then upload it as a new branch to my repo.

The most important issue is probably getting the rebase correct ( make sure you have everything backed up first, because it is possible to screw up rebase in some pretty confusing ways ). When you say the rebase "gave a lot of errors", do you mean errors or merge conflicts? If you've been working on main, and you attempt to naively rebase onto 1.4_maintenance, you're going to be trying to carry all the code from main back to 1.4_maintenance, which will generate piles of merge conflicts. You're going to want to something like `rebase -i`, and select just your commits, and make sure you don't bring along all the unrelated changes from main. If you're still getting merge conflicts, it's possible that there are a few conflicts in the files you've actually changes, but those should be pretty straightforward to address.

If you're getting actual errors though, rather than merge conflicts, then it depends entirely what those errors are.

BrianHanke commented:
@danieldresser Here's what I tried to do:

1. I renamed my main branch to cycles_ui_updates.
2. Then I ran (using a guide I found): 

```
git clone https://github.com/BrianHanke/gaffer.git
cd gaffer
git checkout cycles_ui_updates
git remote add upstream https://github.com/GafferHQ/gaffer.git
git pull --rebase upstream 1.4_maintenance
```
That's when it started telling me there were things it couldn't do with files I didn't recognize. It suggested I rebase --skip which I did, but then it complained about other files that had unresolved conflicts or something like that. 

Anyway, I've now got both main and cycles_ui_updates on my fork and both have the correct updated files. I've also got a local backup!

danieldresser-ie commented:
The first 4 commands there look good. You shouldn't need the --rebase flag when pulling 1.4_maintenance though - 1.4_maintenance is only going to receive changes from upstream, so there shouldn't be any merging required. What you do need to do is checkout a branch for 1.4_maintenance though. I think maybe what this command is doing is trying to rebase 1.4_maintenance on top of your current branch, instead of rebasing your branch on top of maintenance, it makes sense that could cause some problems.

I would usually do
`git fetch upstream` ( to make sure it knows about the 1.4_maintenance branch from upstream, not needed if it already knows about it )
`git checkout 1.4_maintenance` ( switch to the maintenance branch, copying it from upstream if it doesn't already exist )
`git pull upstream 1.4_maintenance` ( grabs an up to date version of the branch from upstream, if it's been modified since you last did this )

Then once you've got a local copy of the 1.4_maintenance branch, you go back to your working branch, and do a rebase:

`git checkout cycles_ui_updates`
`git rebase -i 1.4_maintenance`

That should give you a list of all the commits where your current branch has diverged from maintenance. In this case, you should see way too many commits, because you've got everything from main.

BrianHanke commented:
Ok, I think it might have worked, it told me it rebased to 1.4_maintenance successfully. Is there a way to confirm it's suitable on your end? I still need to update Changes.md and squash the commits.

danieldresser-ie commented:
I'm going to try reopening this PR and seeing if I can repoint it to your new branch.

danieldresser-ie commented:
Oops, looks like Github won't let us change the head branch on a PR. So you'll need to make a new PR with your new branch, based on 1.4_maintenance.

BrianHanke commented:
Ok, I don't think I did it right anyway, there were a million 1.5 files. I just set up a new branch, cycles_ui_final that is 100% based on 1.4_maintenence and only has my new files. I'll try to make the PR from there. Sorry for my GitHub noobishness!

danieldresser-ie commented:
No worries - there are unfortunately quite a few hurdles to starting to contribute to Gaffer - I'm glad to see that you seem to be getting through them!
